### PR TITLE
Feat surf detection

### PR DESCRIPTION
**Pull request by solarbeam**
Add SURF detection functionality

### CODE REVIEW

File: TOPBanga/Detection/SURF.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+#if !__IOS__
+using Emgu.CV.Cuda;
+#endif
+using Emgu.CV.XFeatures2D;
+
+namespace TOPBanga.Detection
+{
+    public static class DrawMatches
+    {
+        public static void FindMatch(Mat modelImage, Mat observedImage, out long matchTime, out VectorOfKeyPoint modelKeyPoints, out VectorOfKeyPoint observedKeyPoints, VectorOfVectorOfDMatch matches, out Mat mask, out Mat homography)
```
Comment (pauliuskalv) [inline TOPBanga/Detection/SURF.cs]: Can you explain the function parameters ( what parameters impact which calculations )

File: TOPBanga/Detection/SURF.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                using (CudaBFMatcher matcher = new CudaBFMatcher(DistanceType.L2))
+                {
+                    surfCuda.DownloadKeypoints(gpuModelKeyPoints, modelKeyPoints);
+                    watch = Stopwatch.StartNew();
+
+                    // extract features from the observed image
+                    using (GpuMat gpuObservedImage = new GpuMat(observedImage))
+                    using (GpuMat gpuObservedKeyPoints = surfCuda.DetectKeyPointsRaw(gpuObservedImage, null))
+                    using (GpuMat gpuObservedDescriptors = surfCuda.ComputeDescriptorsRaw(gpuObservedImage, null, gpuObservedKeyPoints))
+                    //using (GpuMat tmp = new GpuMat())
```
Comment (pauliuskalv) [inline TOPBanga/Detection/SURF.cs]: Remove commented code

File: TOPBanga/SURF_Form.Designer.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            this.Load += new System.EventHandler(this.SURF_Form_Load);
+            ((System.ComponentModel.ISupportInitialize)(this.Webcam_Picture)).EndInit();
+            ((System.ComponentModel.ISupportInitialize)(this.Capture_Picture)).EndInit();
+            this.ResumeLayout(false);
+
+        }
+
+        #endregion
+
+        private System.Windows.Forms.Button button1;
```
Comment (mrzilvis) [inline TOPBanga/SURF_Form.Designer.cs]: This field is never used. Maybe remove it.

File: TOPBanga/SURF_Form.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        private void SURF_Form_Load(object sender, EventArgs e)
+        {
+            this.webcam = new VideoCapture();
+            this.webcam_frame_timer = new System.Timers.Timer();
+            this.webcam_frame_timer.Interval = 30;
+            this.webcam_frame_timer.Elapsed += new ElapsedEventHandler(frame_tick);
+            this.webcam_frame_timer.Start();
+        }
+
+        public void frame_tick(object o, EventArgs e)
```
Comment (mrzilvis) [inline TOPBanga/SURF_Form.cs]: change from frame_tick to Frame_tick

File: TOPBanga/SURF_Form.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        public SURF_Form()
+        {
+            InitializeComponent();
+        }
+
+        private void SURF_Form_Load(object sender, EventArgs e)
+        {
+            this.webcam = new VideoCapture();
+            this.webcam_frame_timer = new System.Timers.Timer();
+            this.webcam_frame_timer.Interval = 30;
```
Comment (mrzilvis) [inline TOPBanga/SURF_Form.cs]: This interval is hardcoded to 30ms, why is it 30ms, User should be able to change the interval in the form.

**Comment by solarbeam** [comment]
I have made changes to the code, as per reviewers' request. Please check out the code again.

File: TOPBanga/Detection/SURF.cs (hunk 1 above)
Comment (solarbeam) [reply TOPBanga/Detection/SURF.cs]: I have added an explanation to the function.

File: TOPBanga/Detection/SURF.cs (hunk 2 above)
Comment (solarbeam) [reply TOPBanga/Detection/SURF.cs]: Removed.

File: TOPBanga/SURF_Form.Designer.cs (hunk 3 above)
Comment (solarbeam) [reply TOPBanga/SURF_Form.Designer.cs]: Fixed in most recent commit.

File: TOPBanga/Detection/SURF.cs (hunk 1 above)
Comment (pauliuskalv) [reply TOPBanga/Detection/SURF.cs]: I think that the explanation is still not enough. However, what do you think @mrzilvis ?

File: TOPBanga/SURF_Form.cs (hunk 5 above)
Comment (solarbeam) [reply TOPBanga/SURF_Form.cs]: Fixed in the newest commit.

File: TOPBanga/SURF_Form.cs (hunk 4 above)
Comment (solarbeam) [reply TOPBanga/SURF_Form.cs]: Done

File: TOPBanga/Detection/SURF.cs (hunk 1 above)
Comment (solarbeam) [reply TOPBanga/Detection/SURF.cs]: Added more explanation

Comment (solarbeam) [reply TOPBanga/Detection/SURF.cs]: mrzilvis agreed

**Pull request merged**